test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,365 @@
+
+    local ok = g.server_1:exec(function()
+        local ok, _ = pcall(box.ctl.promote)
+        box.error.injection.set('ERRINJ_BOX_RAFT_SYNCHRO_QUEUE_DELAY', false)
+        return ok
+    end)
+    luatest.assert(ok, 'error while promoting raft leader')
+end
+
+-- Test that box_demote will return 0
+-- if server is already follower.
+g.before_test('test_follower_demote', function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+    make_follower(g.server_1)
+end)
+
+g.test_follower_demote = function(g)
+    local ok, _ = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, 'error while demoting follower')
+end
+
+-- Test that box_promote will return box.error.UNSUPPORTED
+-- when called while in promote already
+g.test_simultaneous_promote = function(g)
+    local ok, err = g.server_1:exec(function()
+        require('fiber').create(box.ctl.promote)
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.UNSUPPORTED,
+        'error while promoting while in promote')
+end
+
+-- Test that box_demote will return box.error.UNSUPPORTED
+-- when called while in promote already
+g.test_simultaneous_demote = function(g)
+    local ok, err = g.server_1:exec(function()
+        require('fiber').create(box.ctl.promote)
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.UNSUPPORTED,
+        'error while demoting while in promote')
+end
+
+-- Test that box_promote will return box.error.UNSUPPORTED
+-- when trying to promote voter
+g.before_test('test_voter_promote', function(g)
+    g.server_1:box_config({election_mode = 'voter'})
+end)
+
+g.test_voter_promote = function(g)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.UNSUPPORTED,
+        'error while promoting voter')
+end
+
+-- Test interfering promotion for election_mode = 'off'
+-- while in WAL delay
+g.test_wal_interfering_promote = function(g)
+    local wal_write_count = get_wal_write_count(g.server_1)
+    local raft_term = get_raft_term(g.server_1)
+
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_WAL_DELAY', true)
+    end)
+
+    g.server_2:exec(function()
+        box.ctl.promote()
+    end)
+    g.server_2:wait_synchro_queue_owner()
+    wait_promote(g.server_1, wal_write_count, raft_term)
+
+    local ok, err = g.server_1:exec(function()
+        local f = require('fiber').new(box.ctl.promote); f:set_joinable(true)
+        box.error.injection.set('ERRINJ_WAL_DELAY', false)
+        return f:join()
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'interfering promote not handled')
+end
+
+-- Test interfering promotion for election_mode = 'off'
+-- while in txn_limbo delay
+g.test_limbo_empty_interfering_promote = function(g)
```

Comment:
Same. Let's unify this test with `test_limbo_empty_interfering_demote`

Reply:
This one was broken before. It was doing box.ctl.promote in both tests, and the demote branch wasn't covered. I've fixed it. I've ran this test a lot and sometimes (less than once per 1000 runs) it fails on an assert. I am investigating this problem.

```
TAP version 13
1..1
# Started on Mon Jan 10 18:23:28 2022
# Starting group: gh-6033-box-promote-demote
server_1 | tarantool: /home/boris/dev/github/tarantool/src/box/box.cc:1810: void box_issue_demote(uint32_t, int64_t): Assertion `box_raft()->volatile_term == box_raft()->term' failed.
not ok 1        gh-6033-box-promote-demote.test_limbo_empty_interfering_demote
#   Peer closed
#   stack traceback:
#       builtin/box/net_box.lua:676: in function '_request'
#       builtin/box/net_box.lua:728: in function 'eval'
#       ...v/github/tarantool/test/replication-luatest/wip_test.lua:106: in function 'gh-6033-box-promote-demote.test_limbo_empty_interfering_demote'
#       ...
#       [C]: in function 'xpcall'
# Ran 1 tests in 2.528 seconds, 0 successes, 1 error
```